src/hotspot/share/gc/g1/heapRegion.inline.hpp

```diff
@@ -141,7 +141,12 @@ inline bool HeapRegion::block_is_obj(const HeapWord* p) const {
     assert(is_continues_humongous(), "This case can only happen for humongous regions");
     return (p == humongous_start_region()->bottom());
   }
-  if (ClassUnloadingWithConcurrentMark) {
+  // In full gc, we might have skipped compacting some heap regions with high live ratio,
+  // for objs in these regions, the corresponding class info might have been unloaded if
+  // they're not marked in the full gc.
+  // So, only when ClassUnloading is false, it's safe to tell an obj is indeed an obj when
+  // it's under the top of the region, otherwise we have to go to the slow path below.
```

Comment:
This now only talks about the Full GC case and previously we only needed this when unloading during a concurrent cycle. I would like this case to be mentioned as well. Maybe something like:
```suggestion
  // When class unloading is enabled it is not safe to only consider top() to conclude if the 
  // given pointer is a valid object. The situation can occur both for class unloading in a 
  // Full GC and during a concurrent cycle. 
  // During a Full GC regions can be excluded from compaction due to high live ration, and
  // because of this there can be stale objects for unloaded classes left in these regions.
  // During a concurrent cycle class unloading is done after marking is complete and objects
  // for the unloaded classes will be stale until the regions are collected.
```

Feel free to change the wording.

Reply:
Thanks for refining the comments.